sqlglot/optimizer/scope.py

```diff
@@ -511,82 +535,99 @@ def _traverse_union(scope):
-    _add_table_sources(scope)
-    _traverse_subqueries(scope)
-
-
-def _traverse_derived_tables(derived_tables, scope, scope_type):
+def _traverse_ctes(scope):
     sources = {}
-    is_cte = scope_type == ScopeType.CTE
 
-    for derived_table in derived_tables:
+    for cte in scope.ctes:
         recursive_scope = None
 
         # if the scope is a recursive cte, it must be in the form of
         # base_case UNION recursive. thus the recursive scope is the first
         # section of the union.
-        if is_cte and scope.expression.args["with"].recursive:
-            union = derived_table.this
+        if scope.expression.args["with"].recursive:
+            union = cte.this
 
             if isinstance(union, exp.Union):
                 recursive_scope = scope.branch(union.this, scope_type=ScopeType.CTE)
 
         for child_scope in _traverse_scope(
             scope.branch(
-                derived_table if isinstance(derived_table, exp.UDTF) else derived_table.this,
-                chain_sources=sources if scope_type == ScopeType.CTE else None,
-                outer_column_list=derived_table.alias_column_names,
-                scope_type=ScopeType.UDTF if isinstance(derived_table, exp.UDTF) else scope_type,
+                cte.this,
+                chain_sources=sources,
+                outer_column_list=cte.alias_column_names,
+                scope_type=ScopeType.CTE,
             )
         ):
             yield child_scope
 
-            # Tables without aliases will be set as ""
-            # This shouldn't be a problem once qualify_columns runs, as it adds aliases on everything.
-            # Until then, this means that only a single, unaliased derived table is allowed (rather,
-            # the latest one wins.
-            alias = derived_table.alias
+            alias = cte.alias
             sources[alias] = child_scope
 
             if recursive_scope:
                 child_scope.add_source(alias, recursive_scope)
 
         # append the final child_scope yielded
-        if is_cte:
-            scope.cte_scopes.append(child_scope)
-        else:
-            scope.derived_table_scopes.append(child_scope)
+        scope.cte_scopes.append(child_scope)
 
     scope.sources.update(sources)
 
 
-def _add_table_sources(scope):
+def _traverse_tables(scope):
     sources = {}
-    for table in scope.tables:
-        table_name = table.name
 
-        if table.alias:
-            source_name = table.alias
-        else:
-            source_name = table_name
+    # Traverse FROMs, JOINs, and LATERALs in the order they are defined
+    expressions = []
+    from_ = scope.expression.args.get("from")
+    if from_:
+        expressions.extend(from_.expressions)
 
-        if table_name in scope.sources:
-            # This is a reference to a parent source (e.g. a CTE), not an actual table.
-            scope.sources[source_name] = scope.sources[table_name]
+    for join in scope.expression.args.get("joins") or []:
```

Comment:
This appends `join.this`, and I think `or` is safer, in case "joins" is None.

We could do
```python
expressions.extend(j.this for j in scope.expression.args.get("joins") or [])
```

But I think that's harder to read.